enhancements/machine-config/pin-and-pre-load-images.md

```diff
@@ -0,0 +1,237 @@
+more consistent upgrade time in those environments. That is important when
+scheduling upgrades into maintenance windows, even if the upgrade might not
+otherwise fail.
+
+### User Stories
+
+#### Pre-load and pin upgrade images
+
+As the administrator of a cluster that has a low bandwidth and/or unreliable
+connection to an image registry server I want to pin and pre-load all the
+images required for the upgrade in advance, so that when I decide to actually
+perform the upgrade there will be no need to contact that slow and/or
+unreliable registry server and the upgrade will successfully complete in a
+predictable time.
+
+#### Pre-load and pin application images
+
+As the administrator of a cluster that has a low bandwidth and/or unreliable
+connection to an image registry server I want to pin and pre-load the images
+required by my application in advance, so that when I decide to actually deploy
+it there will be no need to contact that slow and/or unreliable registry server
+and my application will successfully deploy in a predictable time.
+
+### Goals
+
+Provide a mechanism that cluster administrators can use to pin and pre-load
+container images.
+
+### Non-Goals
+
+We wish to use the mechanism described in this enhancement to orchestrate
+upgrades without a registry server. But that orchestration is not a goal
+of this enhancement; it will be part of a separate enhancement based on
+parts of https://github.com/openshift/enhancements/pull/1432.
+
+## Proposal
+
+### Workflow Description
+
+1. The administrator of a cluster uses the `ContainerRuntimeConfig` object to
+request that a set of container images are pinned and pre-loaded:
+
+    ```yaml
+    apiVersion: machineconfiguration.openshift.io/v1
+    kind: ContainerRuntimeConfig
+    metadata:
+      name: ...
+    spec:
+      containerRuntimeConfig:
+        pinnedImages:
+        - quay.io/openshift-release-dev/ocp-release@sha256:...
+        - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+        - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+        ...
+    ```
+
+1. The machine config operators ensures that all the images are pinned and
+pulled in all the nodes of the cluster.
+
+### API Extensions
+
+There are no new object kinds introduced by this enhancement, but new fields
+will be added to existing `ContainerRuntimeConfig` objects.
+
+The new fields for the `ContainerRuntimeConfig` object are defined in detail in
+https://github.com/openshift/machine-config-operator/pull/3839.
+
+### Implementation Details/Notes/Constraints
+
+Starting with version 4.14 of OpenShift CRI-O will have the capability to pin
+certain images (see [this](https://github.com/cri-o/cri-o/pull/6862) pull
+request for details). That capability will be used to pin all the images
+required for the upgrade, so that they aren't garbage collected by kubelet and
+CRI-O.
+
+The changes to pin the images will be done in a `/etc/crio/crio.conf.d/pin.conf`
```

Comment:
Admins could override that file which would cause a race. Do we need something like "multiple config-dirs" in CRI-O or a reserved namespace in the MCO?

Reply:
Maybe this file could be named after the `ContainerRuntimeConfig` UUID (or the `PinnedImagesSet` UUID, if we agree to introduce that). Could be something like this:

```
pinned-images-be8545cb-dd0e-48cd-8e4f-687babc5b1b2.conf
```

Less likely to conflict with files created by admins, but not totally impossible.

Reply:
Yeah I think that's better

Reply:
OK, added that in the latest patch.